py/picca/io.py

```diff
@@ -261,42 +260,111 @@ def read_data(in_dir,drq,mode,zmin = 2.1,zmax = 3.5,nspec=None,log=None,keep_bal
+            sys.exit(1)
+        if len(fi) == 0:
+            print("ERROR: can't find required spAll file in {}".format(in_dir))
+            print("ERROR: try runnint with --best-obs option (but you will lose reobservations)")
+            sys.exit(1)
+
+        spAll = fitsio.FITS(fi[0])
+        print("INFO: reading spAll from {}".format(fi[0]))
+        thid_spall = spAll[1]["THING_ID"][:]
+        plate_spall = spAll[1]["PLATE"][:]
+        mjd_spall = spAll[1]["MJD"][:]
+        fid_spall = spAll[1]["FIBERID"][:]
+        qual_spall = spAll[1]["PLATEQUALITY"][:]
+        zwarn_spall = spAll[1]["ZWARNING"][:]
+
+        w = sp.in1d(thid_spall, thid) & (qual_spall == b"good")
+        ## Removing spectra with the following ZWARNING bits set:
+        ## SKY, LITTLE_COVERAGE, UNPLUGGED, BAD_TARGET, NODATA
+        ## https://www.sdss.org/dr14/algorithms/bitmasks/#ZWARNING
+        for zwarnbit in [0,1,7,8,9]:
+            w &= (zwarn_spall&2**zwarnbit)==0
+        print("INFO: # unique objs: ",len(thid))
+        print("INFO: # spectra: ",w.sum())
+        thid = thid_spall[w]
+        plate = plate_spall[w]
+        mjd = mjd_spall[w]
+        fid = fid_spall[w]
+        spAll.close()
+
+    ## to simplify, use a list of all metadata
+    allmeta = []
+    t_list = []
+    for t,p,m,f in zip(thid,plate,mjd,fid):
+        t_list.append(t)
+        r,d,z = drq_dict[t]
+        meta = metadata()
+        meta.thid = t
+        meta.ra = r
+        meta.dec = d
+        meta.zqso = z
+        meta.plate = p
+        meta.mjd = m
+        meta.fid = f
+        meta.order = order
+        allmeta.append(meta)
+
     pix_data = []
-    for t,r,d,z,p,m,f in zip(thid,ra,dec,zqso,plate,mjd,fid):
-        try:
-            fid = str(f)
-            if f<10:
-                fid='0'+fid
-            if f<100:
-                fid = '0'+fid
-            if f<1000:
-                fid = '0'+fid
-            fin = in_dir + "/{}/{}-{}-{}-{}.fits".format(p,mode,p,m,fid)
-            h = fitsio.FITS(fin)
-        except IOError:
-            log.write("error reading {}\n".format(fin))
-            continue
+    thids = {}
 
-        log.write("{} read\n".format(fin))
-        ll = h[1]["loglam"][:]
-        fl = h[1]["flux"][:]
-        iv = h[1]["ivar"][:]*(h[1]["and_mask"][:]==0)
-
-        if(pk1d is not None) :
-            # compute difference between exposure
-            diff = exp_diff(h,ll)
-            # compute spectral resolution
-            wdisp =  h[1]["wdisp"][:]
-            reso = spectral_resolution(wdisp,True,f,ll)
-            d = forest(ll,fl,iv, t, r, d, z, p, m, f,order,diff,reso)
-        else :
-            d = forest(ll,fl,iv, t, r, d, z, p, m, f,order)
 
-        pix_data.append(d)
-        h.close()
-    return pix_data
+    for meta in allmeta:
+        t = meta.thid
+        if not t in thids:
+            thids[t] = []
+        thids[t].append(meta)
+
+    print("reading {} thids".format(len(thids)))
 
+    for t in t_list:
```

Comment:
this could be a `for t in thids.keys()` or even a `for t,tmeta in thids.items():` followed by `for meta in tmeta:`. Just to be independent of the t_list array which is elsewise not needed anymore

Reply:
also: I think t_list should contain the same tid multiple times, whereas the data in thids is unique for each tid, so I guess this is actually an issue that might lead to double counting spectra.

Reply:
Note that for the set approach you need to change the iterator here to t_set